experimental/incremental/query.go

```diff
@@ -49,24 +49,23 @@ type Query[T any] interface {
 	Execute(*Task) (value T, fatal error)
 }
 
-// ErrCycle is returned by [Resolve] if a cycle occurs during query execution.
-type ErrCycle struct {
-	// The offending cycle. The first and last queries will have the same URL.
-	//
-	// To inspect the concrete types of the cycle members, use [DowncastQuery],
-	// which will automatically unwrap any calls to [AnyQuery].
-	Cycle []*AnyQuery
+// ErrCycle is an error due to cyclic dependencies.
+//
+// When returned by [Resolve] or [Run], this will be *ErrCycle[*AnyQuery].
+type ErrCycle[T any] struct {
```

Comment:
Perhaps move the comment about the concrete type of `T` to the doc comments for `Resolve` and `Run`. That way, reading the function docs is enough to understand the types of returned errors.

We may also want to consider moving this generic type to an internal package and then  then creating statically-typed exported aliases, like `type ErrQueryCycle = internal.ErrCycle[*AnyQuery]` and `type ErrImportCycle = internal.ErrCycle[ast.DeclImport]`.

Reply:
That's not a bad idea, actually.